src/components/modals/Username.js

```diff
@@ -1,125 +1,119 @@
-import React, { PureComponent } from 'react';
+import React, { useRef, useState } from 'react';
 import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
 
 import { Modal, Button } from 'react-bootstrap';
 
 import { generateRandomUsername } from '../../data/username';
 
+import { enableAudio } from '../../store/actions/settingsActions';
+
 export const MAX_USERNAME_LENGTH = 45;
 
-class UsernameModal extends PureComponent {
-  constructor(props) {
-    super(props);
-    this.state = { failMessage: '' };
-  }
+const UsernameModal = ({
+  isVisible,
+  previousUsername,
+  username,
+  onToggleModalVisibility,
+  setUsername,
+  enableAudio,
+}) => {
+  const usernameInput = useRef(null);
 
-  componentDidMount() {
-    this.usernameInput.focus();
-  }
+  const [failMessage, setFailMessage ] = useState("");
 
-  componentDidUpdate(){
-    if (this.props.showModal) {
-      this.usernameInput.focus();
-    }
-  }
+  const onClose = () => {
+    onToggleModalVisibility('username', false);
+  };
 
-  onUsernameKeyPress = (e) => {
-    if (e.which === 13) {
-      this.onSetUsernameClick();
-    }
+  const setRandomUsernameInForm = () => {
+    usernameInput.current.value = generateRandomUsername();
+    usernameInput.current.focus();
   };
 
-  isUsernameValid(username) {
-    if (!username || username.length === 0) {
-      this.setState({ failMessage: 'Must not be empty' });
+  const isUsernameValid = () => {
+    const usernameFromForm = usernameInput.current.value;
+    if (!usernameFromForm || usernameFromForm.length === 0) {
+      setFailMessage("Must not be empty");
       return false;
-    } else if (username.length > MAX_USERNAME_LENGTH) {
-      this.setState({ failMessage: `Length must not exceed ${MAX_USERNAME_LENGTH}` });
+    } else if (usernameFromForm.length > MAX_USERNAME_LENGTH) {
+      setFailMessage(`Length must not exceed ${MAX_USERNAME_LENGTH}`);
       return false;
-    } else {
-      return true;
-    }
-  }
-
-  onSetUsernameClick = (e) => {
-    const username = this.usernameInput.value;
-
-    if (this.isUsernameValid(username)) {
-      this.props.onSetUsername(username);
-      // set the audio to the user's previously selected preference; enable by default
-      let isAudioEnabled = JSON.parse(localStorage.getItem('isAudioEnabled') || 'true');
-      this.props.onSetIsAudioEnabled(isAudioEnabled);
     }
+    setFailMessage("");
```

Comment:
Does this not trigger an error message with an empty body or something? This is right?

Reply:
See line 13, in the old state object, above. The default was an empty string.

I also didn't see a fail message when I shouldn't have in the browser.

It is a moot point, however. In the old code, nothing reset the failMessage if the username was valid, after being invalid. The next thing that happens if we hit this part of the code is we close the modal.

Reply:
Gotcha, thanks.